### ISSUE TITLE
Resolve hotfix differences in Trilinos/Kokkos

### ISSUE THREAD

**Issue by crtrott** [kokkos/kokkos#2491]
Here is a diff from Trilinos if I copy the kokkos in Trilinos into the master branch:

We need to check which of these things we don't cover in develop and address that. 

```
-bash-4.2$ git status
On branch master
Your branch is up-to-date with 'origin/master'.
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   containers/performance_tests/CMakeLists.txt
	modified:   core/perf_test/CMakeLists.txt
	modified:   core/src/Kokkos_Complex.hpp
	modified:   core/src/impl/Kokkos_Core.cpp
	modified:   core/unit_test/TestComplex.hpp

no changes added to commit (use "git add" and/or "git commit -a")
-bash-4.2$ git diff
diff --git a/containers/performance_tests/CMakeLists.txt b/containers/performance_tests/CMakeLists.txt
index 3c6584b..8519435 100644
--- a/containers/performance_tests/CMakeLists.txt
+++ b/containers/performance_tests/CMakeLists.txt
@@ -37,11 +37,12 @@ TRIBITS_ADD_EXECUTABLE(
   SOURCES ${SOURCES}
   COMM serial mpi
   TESTONLYLIBS kokkos_gtest ${TEST_LINK_TARGETS}
+  CATEGORIES BASIC PERFORMANCE
   )
 
 TRIBITS_ADD_TEST(
-  PerformanceTest
-  NAME PerfTestExec
+  PerfTestExec
+  NAME PerformanceTest
   COMM serial mpi
   NUM_MPI_PROCS 1
   CATEGORIES PERFORMANCE
diff --git a/core/perf_test/CMakeLists.txt b/core/perf_test/CMakeLists.txt
index d92462a..e92958b 100644
--- a/core/perf_test/CMakeLists.txt
+++ b/core/perf_test/CMakeLists.txt
@@ -32,11 +32,12 @@ TRIBITS_ADD_EXECUTABLE(
   SOURCES ${SOURCES}
   COMM serial mpi
   TESTONLYLIBS kokkos_gtest ${TEST_LINK_TARGETS}
+  CATEGORIES BASIC PERFORMANCE
   )
 
 TRIBITS_ADD_TEST(
-  PerfTest
-  NAME PerfTestExec
+  PerfTestExec
+  NAME PerfTest
   COMM serial mpi
   NUM_MPI_PROCS 1
   CATEGORIES PERFORMANCE
@@ -48,6 +49,7 @@ TRIBITS_ADD_EXECUTABLE(
   SOURCES test_taskdag.cpp
   COMM serial mpi
   TESTONLYLIBS kokkos_gtest ${TEST_LINK_TARGETS}
+  CATEGORIES BASIC PERFORMANCE
   )
 
 TRIBITS_ADD_TEST(
diff --git a/core/src/Kokkos_Complex.hpp b/core/src/Kokkos_Complex.hpp
index a3ada5d..5e83bd1 100644
--- a/core/src/Kokkos_Complex.hpp
+++ b/core/src/Kokkos_Complex.hpp
@@ -812,11 +812,11 @@ std::ostream& operator << (std::ostream& os, const complex<RealType>& x) {
 }
 
 template<class RealType>
-std::ostream& operator >> (std::ostream& os, complex<RealType>& x) {
+std::istream& operator >> (std::istream& is, complex<RealType>& x) {
   std::complex<RealType> x_std;
-  os >> x_std;
+  is >> x_std;
   x = x_std; // only assigns on success of above
-  return os;
+  return is;
 }
 
 
diff --git a/core/src/impl/Kokkos_Core.cpp b/core/src/impl/Kokkos_Core.cpp
index 0d472e9..d610a29 100644
--- a/core/src/impl/Kokkos_Core.cpp
+++ b/core/src/impl/Kokkos_Core.cpp
@@ -50,6 +50,7 @@
 #include <cstdlib>
 #include <stack>
 #include <cerrno>
+#include <unistd.h>
   CATEGORIES PERFORMANCE
@@ -48,6 +49,7 @@ TRIBITS_ADD_EXECUTABLE(
   SOURCES test_taskdag.cpp
   COMM serial mpi
   TESTONLYLIBS kokkos_gtest ${TEST_LINK_TARGETS}
+  CATEGORIES BASIC PERFORMANCE
   )
 
 TRIBITS_ADD_TEST(
diff --git a/core/src/Kokkos_Complex.hpp b/core/src/Kokkos_Complex.hpp
index a3ada5d..5e83bd1 100644
--- a/core/src/Kokkos_Complex.hpp
+++ b/core/src/Kokkos_Complex.hpp
@@ -812,11 +812,11 @@ std::ostream& operator << (std::ostream& os, const complex<RealType>& x) {
 }
 
 template<class RealType>
-std::ostream& operator >> (std::ostream& os, complex<RealType>& x) {
+std::istream& operator >> (std::istream& is, complex<RealType>& x) {
   std::complex<RealType> x_std;
-  os >> x_std;
+  is >> x_std;
   x = x_std; // only assigns on success of above
-  return os;
+  return is;
 }
 
 
diff --git a/core/src/impl/Kokkos_Core.cpp b/core/src/impl/Kokkos_Core.cpp
index 0d472e9..d610a29 100644
--- a/core/src/impl/Kokkos_Core.cpp
+++ b/core/src/impl/Kokkos_Core.cpp
@@ -50,6 +50,7 @@
 #include <cstdlib>
 #include <stack>
 #include <cerrno>
+#include <unistd.h>
 
 //----------------------------------------------------------------------------
 
@@ -605,18 +606,33 @@ void initialize(int& narg, char* arg[])
         else
             device = env_device;
     }
+    auto env_rdevices_str = std::getenv("KOKKOS_RAND_DEVICES");
     auto env_ndevices_str = std::getenv("KOKKOS_NUM_DEVICES");
-    if (env_ndevices_str!=nullptr) {
+    if (env_ndevices_str!=nullptr || env_rdevices_str!=nullptr) {
         errno = 0;
-        auto env_ndevices = std::strtol(env_ndevices_str,&endptr,10);
-        if (endptr== env_ndevices_str) 
+        if (env_ndevices_str!=nullptr && env_rdevices_str!=nullptr) {
+          Impl::throw_runtime_exception("Error: cannot specify KOKKOS_NUM_DEVICES and KOKKOS_RAND_DEVICES. Raised by Kokkos::initialize(int narg, char* argc[]).");
+        }
+        int rdevices=-1;
+        if (env_ndevices_str!=nullptr) {
+          auto env_ndevices = std::strtol(env_ndevices_str,&endptr,10);
+          if (endptr== env_ndevices_str)
             Impl::throw_runtime_exception("Error: cannot convert KOKKOS_NUM_DEVICES to an integer. Raised by Kokkos::initialize(int narg, char* argc[]).");
-        if (errno == ERANGE)
+          if (errno == ERANGE)
             Impl::throw_runtime_exception("Error: KOKKOS_NUM_DEVICES out of range of representable values by an integer. Raised by Kokkos::initialize(int narg, char* argc[]).");
-        if ((ndevices != -1)&&(env_ndevices!=ndevices))
+          if ((ndevices != -1)&&(env_ndevices!=ndevices))
             Impl::throw_runtime_exception("Error: expecting a match between --kokkos-ndevices and KOKKOS_NUM_DEVICES if both are set. Raised by Kokkos::initialize(int narg, char* argc[]).");
-        else
+          else
             ndevices = env_ndevices;
+        } else { //you set KOKKOS_RAND_DEVICES
+          auto env_rdevices = std::strtol(env_rdevices_str,&endptr,10);
 
 //----------------------------------------------------------------------------
 
@@ -605,18 +606,33 @@ void initialize(int& narg, char* arg[])
         else
             device = env_device;
     }
+    auto env_rdevices_str = std::getenv("KOKKOS_RAND_DEVICES");
     auto env_ndevices_str = std::getenv("KOKKOS_NUM_DEVICES");
-    if (env_ndevices_str!=nullptr) {
+    if (env_ndevices_str!=nullptr || env_rdevices_str!=nullptr) {
         errno = 0;
-        auto env_ndevices = std::strtol(env_ndevices_str,&endptr,10);
-        if (endptr== env_ndevices_str) 
+        if (env_ndevices_str!=nullptr && env_rdevices_str!=nullptr) {
+          Impl::throw_runtime_exception("Error: cannot specify KOKKOS_NUM_DEVICES and KOKKOS_RAND_DEVICES. Raised by Kokkos::initialize(int narg,
 char* argc[]).");
+        }
+        int rdevices=-1;
+        if (env_ndevices_str!=nullptr) {
+          auto env_ndevices = std::strtol(env_ndevices_str,&endptr,10);
+          if (endptr== env_ndevices_str)
             Impl::throw_runtime_exception("Error: cannot convert KOKKOS_NUM_DEVICES to an integer. Raised by Kokkos::initialize(int narg, char* a
rgc[]).");
-        if (errno == ERANGE)
+          if (errno == ERANGE)
             Impl::throw_runtime_exception("Error: KOKKOS_NUM_DEVICES out of range of representable values by an integer. Raised by Kokkos::initia
lize(int narg, char* argc[]).");
-        if ((ndevices != -1)&&(env_ndevices!=ndevices))
+          if ((ndevices != -1)&&(env_ndevices!=ndevices))
             Impl::throw_runtime_exception("Error: expecting a match between --kokkos-ndevices and KOKKOS_NUM_DEVICES if both are set. Raised by K
okkos::initialize(int narg, char* argc[]).");
-        else
+          else
             ndevices = env_ndevices;
+        } else { //you set KOKKOS_RAND_DEVICES
+          auto env_rdevices = std::strtol(env_rdevices_str,&endptr,10);
+          if (endptr== env_ndevices_str)
+            Impl::throw_runtime_exception("Error: cannot convert KOKKOS_RAND_DEVICES to an integer. Raised by Kokkos::initialize(int narg, char* argc[]).");
+          if (errno == ERANGE)
+            Impl::throw_runtime_exception("Error: KOKKOS_RAND_DEVICES out of range of representable values by an integer. Raised by Kokkos::initialize(int narg, char* argc[]).");
+          else
+            rdevices = env_rdevices;
+        }
         //Skip device
         auto env_skip_device_str = std::getenv("KOKKOS_SKIP_DEVICE");
         if (env_skip_device_str!=nullptr) {
@@ -631,6 +647,17 @@ void initialize(int& narg, char* arg[])
             else
                 skip_device = env_skip_device;
         }
+        if (rdevices>0) {
+          if (skip_device > 0 && rdevices == 1)
+            Impl::throw_runtime_exception("Error: cannot KOKKOS_SKIP_DEVICE the only KOKKOS_RAND_DEVICE.Raised by Kokkos::initialize(int narg, char* argc[]).");
+
+          std::srand(getpid());
+          while (device < 0 ) {
+            int test_device = std::rand()%rdevices;
+            if (test_device != skip_device)
+              device=test_device;
+          }
+        }
     }
     char * env_disablewarnings_str = std::getenv("KOKKOS_DISABLE_WARNINGS");
     if (env_disablewarnings_str!=nullptr) {
diff --git a/core/unit_test/TestComplex.hpp b/core/unit_test/TestComplex.hpp
index f4343df..6ec0739 100644
--- a/core/unit_test/TestComplex.hpp
+++ b/core/unit_test/TestComplex.hpp
@@ -43,6 +43,7 @@
 
 #include<Kokkos_Core.hpp>
 #include<cstdio>
+#include <sstream>
 
 namespace Test {
 
@@ -254,10 +255,29 @@ struct TestComplexSpecialFunctions {
   }
 };
 
+void testComplexIO() {
+  Kokkos::complex<double> z = {3.14, 1.41};
+  std::stringstream ss;
+  ss << z;
+  ASSERT_EQ(ss.str(), "(3.14,1.41)");
+
+  ss.str("1 (2) (3,4)");
+  ss.clear();
+  ss >> z;
+  ASSERT_EQ(z, (Kokkos::complex<double>{1, 0}));
+  ss >> z;
+  ASSERT_EQ(z, (Kokkos::complex<double>{2, 0}));
+  ss >> z;
+  ASSERT_EQ(z, (Kokkos::complex<double>{3, 4}));
+}
+
 TEST_F(TEST_CATEGORY, complex_special_funtions) {
   TestComplexSpecialFunctions<TEST_EXECSPACE> test;
   test.testit();
 }
+
+TEST_F(TEST_CATEGORY, complex_io) { testComplexIO(); }
+
 } // namespace Test
```

**Comment by ndellingwood** [comment]
References to issues regarding the discrepancies:
The Kokkos_Core changes look like those discussed in #2254 regarding the ability to pick a device at random (intended to improve testing performance) with a PR merged into Trilinos by PR trilinos/Trilinos#5799, though never added to Kokkos.
The Complex fixes for the >> operator with unit tests were addressed in PR #2320, with matching trilinos PR trilinos/Trilinos#5899

**Comment by masterleinad** [comment]
Addressed in #2498.
